# Filmijob Capstone Project: Interview Question Scoring

In [1]:
import pandas as pd
import re

## 1. EDA (Exploratory Data Analysis)

### 1.1 Importing the Dataset and checking the data

In [2]:
df=pd.read_csv("Dataset.csv")
df

title  \
0          Collection Specialist   
1                Billing Analyst   
2                 Tax Accountant   
3    Accounts Receivable Manager   
4                Cost Accountant   
..                           ...   
278                Sales Advisor   
279                 Retail Buyer   
280             Category Manager   
281                 Merchandiser   
282                Store Manager   

                                       job description  \
0    <h4>Job brief</h4><p>We are looking for a comp...   
1    <h4>Job brief</h4><p>We are looking for a Bill...   
2    <h4>Job brief</h4><p>We are looking for a Tax ...   
3    <h4>Job brief</h4><p>We are looking for an Acc...   
4    <h4>Job brief</h4><p>We are looking for a Cost...   
..                                                 ...   
278  <h4>Job brief</h4><p><span style="font-weight:...   
279  <h4>Job brief</h4><p><span style="font-weight:...   
280  <h4>Job brief</h4><p><span style="font-weight:...   
281  <h4>Job brief</h4><p>We are looking for an exp...   
282  <h4>Job brief</h4><p>We are looking for a resu...   

                                             questions  
0    ['A debtor has missed six months of payments. ...  
1    ['How would you create a report on debts from ...  
2    ['What would you do if you received a B-Notice...  
3    ['What would be the first thing you’d check if...  
4    ['If you could use only Excel, how would you o...  
..                                                 ...  
278  ['Imagine a customer asks you a question about...  
279  ['If I asked you to eliminate one line of our ...  
280  ['Here’s a spreadsheet with data. Can you spot...  
281  ['As a merchandiser, what do you think is the ...  
282  ['Who are our main competitors?', 'What tasks,...  

[283 rows x 3 columns]

### 1.2 Counting the number of keywords and checking for repeated keywords

In [3]:
opened_file= open("keywords.txt")
from csv import reader
read_file = reader(opened_file)
keywords_data = list(read_file)

In [4]:
keywords_data

[['teaching'],
 ['pharmaceutical'],
 ['ERP'],
 ['fulfillment'],
 ['positioning'],
 ['tax'],
 ['service delivery'],
 ['investigate'],
 ['editorial'],
 ['account management'],
 ['business process'],
 ['valid drivers license'],
 ['electronics'],
 ['PR'],
 ['public relations'],
 ['Javascript'],
 ['assembly'],
 ['digital marketing'],
 ['Linux'],
 ['Facebook'],
 ['spreadsheets'],
 ['recruit'],
 ['proposal'],
 ['SharePoint'],
 ['data entry'],
 ['hotel'],
 ['ordering'],
 ['branding'],
 ['life cycle'],
 ['real estate'],
 ['relationship management'],
 ['researching'],
 ['process improvements'],
 ['chemistry'],
 ['SaaS'],
 ['CAD'],
 ['sales experience'],
 ['mathematics'],
 ['customer-facing'],
 ['audio'],
 ['project management skills'],
 ['six sigma'],
 ['hospitality'],
 ['mechanical engineering'],
 ['auditing'],
 ['employee relations'],
 ['android'],
 ['security clearance'],
 ['licensing'],
 ['Adobe'],
 ['fundraising'],
 ['repairs'],
 ['ISO'],
 ['market research'],
 ['warehouse'],
 ['business st

In [5]:
len(keywords_data)

1410

In [6]:
keyword_list=[]
for keyword in keywords_data:
    keyword_list.append(keyword[0])
keyword_list
len(keyword_list)

1410

In [7]:
all_keywords_set=set(keyword_list)
all_keywords_list=list(all_keywords_set)
len(all_keywords_list)

1054

In [8]:
len(keyword_list)-len(all_keywords_list)

356

There are **356** keyword repetitions

### 1.3 Creating a new column with the number of question per role

In [9]:
df["number_of_questions"]=df["questions"].str.split("[^a-e]', ?'[^a-e]").str.len()
df

title  \
0          Collection Specialist   
1                Billing Analyst   
2                 Tax Accountant   
3    Accounts Receivable Manager   
4                Cost Accountant   
..                           ...   
278                Sales Advisor   
279                 Retail Buyer   
280             Category Manager   
281                 Merchandiser   
282                Store Manager   

                                       job description  \
0    <h4>Job brief</h4><p>We are looking for a comp...   
1    <h4>Job brief</h4><p>We are looking for a Bill...   
2    <h4>Job brief</h4><p>We are looking for a Tax ...   
3    <h4>Job brief</h4><p>We are looking for an Acc...   
4    <h4>Job brief</h4><p>We are looking for a Cost...   
..                                                 ...   
278  <h4>Job brief</h4><p><span style="font-weight:...   
279  <h4>Job brief</h4><p><span style="font-weight:...   
280  <h4>Job brief</h4><p><span style="font-weight:...   
281  <h4>Job brief</h4><p>We are looking for an exp...   
282  <h4>Job brief</h4><p>We are looking for a resu...   

                                             questions  number_of_questions  
0    ['A debtor has missed six months of payments. ...                   16  
1    ['How would you create a report on debts from ...                   14  
2    ['What would you do if you received a B-Notice...                   14  
3    ['What would be the first thing you’d check if...                   17  
4    ['If you could use only Excel, how would you o...                   15  
..                                                 ...                  ...  
278  ['Imagine a customer asks you a question about...                   16  
279  ['If I asked you to eliminate one line of our ...                   19  
280  ['Here’s a spreadsheet with data. Can you spot...                   15  
281  ['As a merchandiser, what do you think is the ...                   15  
282  ['Who are our main competitors?', 'What tasks,...                   14  

[283 rows x 4 columns]

In [10]:
df.describe()

number_of_questions
count           283.000000
mean             15.342756
std               3.106703
min               7.000000
25%              13.000000
50%              15.000000
75%              17.000000
max              29.000000

On **average**, we have **15** questions per role. Also, the role with **less** questions has **7** and the role with the max questions has **29** questions.

### 1.4 Cheking if there are any duplicated roles

In [11]:
all_jobs=df["title"].tolist()
all_jobs_set=set(all_jobs)
all_jobs_list=list(all_jobs_set)
len(all_jobs_list)

283

All the rows have a different role and no role is duplicated

### 1.5 Checking if any of the columns has null values

In [12]:
df.isnull().any()

title                  False
job description        False
questions              False
number_of_questions    False
dtype: bool

No null values on the dataset

### 1.6 Inspecting the total number of questions and questions repetitions

First, we are goint to split the questions in every row of the questions column into a list of lists:

In [13]:
df["questions"]=df["questions"].str.split("[^a-e]', ?'[^a-e]")
df

title  \
0          Collection Specialist   
1                Billing Analyst   
2                 Tax Accountant   
3    Accounts Receivable Manager   
4                Cost Accountant   
..                           ...   
278                Sales Advisor   
279                 Retail Buyer   
280             Category Manager   
281                 Merchandiser   
282                Store Manager   

                                       job description  \
0    <h4>Job brief</h4><p>We are looking for a comp...   
1    <h4>Job brief</h4><p>We are looking for a Bill...   
2    <h4>Job brief</h4><p>We are looking for a Tax ...   
3    <h4>Job brief</h4><p>We are looking for an Acc...   
4    <h4>Job brief</h4><p>We are looking for a Cost...   
..                                                 ...   
278  <h4>Job brief</h4><p><span style="font-weight:...   
279  <h4>Job brief</h4><p><span style="font-weight:...   
280  <h4>Job brief</h4><p><span style="font-weight:...   
281  <h4>Job brief</h4><p>We are looking for an exp...   
282  <h4>Job brief</h4><p>We are looking for a resu...   

                                             questions  number_of_questions  
0    [['A debtor has missed six months of payments....                   16  
1    [['How would you create a report on debts from...                   14  
2    [['What would you do if you received a B-Notic...                   14  
3    [['What would be the first thing you’d check i...                   17  
4    [['If you could use only Excel, how would you ...                   15  
..                                                 ...                  ...  
278  [['Imagine a customer asks you a question abou...                   16  
279  [['If I asked you to eliminate one line of our...                   19  
280  [['Here’s a spreadsheet with data. Can you spo...                   15  
281  [['As a merchandiser, what do you think is the...                   15  
282  [['Who are our main competitors, hat tasks, wh...                   14  

[283 rows x 4 columns]

After having a list of questions for every role, we are going to create a list of all the questions in the dataset:

In [14]:
all_questions=[]
for questions in df["questions"]:
    for question in questions:
        all_questions.append(question)
all_questions

["['A debtor has missed six months of payments. When you call the individual, they bring up financial difficulties and get emotional. How do you handle this",
 'hat payback plan would you recommend for a long-term client whose payments are due to be collected in X months',
 'f you called a client to update them on their payment status and they were aggressive or rude to you, how would you handle it',
 'an you think of a few common excuses that debtors make? How would you respond to each of them',
 'hat collection methods do you know? Which technique do you think would be more effective for our company/clients',
 'ow do you keep historical data for each account',
 'ow often do you follow up with delinquent account holders and what’s your approach when you get in touch',
 'ention three things a Collection Specialist should avoid saying when speaking to a debtor',
 'ow do you keep track of all payments',
 'hat software do you use to increase your productivity',
 'hat do you find interesti

In [15]:
len(all_questions)

4342

There are **4342** total questions

In [16]:
set_allquestions=set(all_questions)
unique_questions=list(set_allquestions)
len(unique_questions)

4141

There are **4141** different questions

In [17]:
num_repeated_questions=len(all_questions)-len(unique_questions)
num_repeated_questions

201

There are **201** question repetitions

Now, we are going to create a frequency table of questions to double check total number and repetitions:

In [18]:
questions_freq={}
for questions in df["questions"]:
    for question in questions:
        if question in questions_freq.keys():
            questions_freq[question]+=1
        else:
            questions_freq[question]=1
            
questions_freq    

{"['A debtor has missed six months of payments. When you call the individual, they bring up financial difficulties and get emotional. How do you handle this": 1,
 'hat payback plan would you recommend for a long-term client whose payments are due to be collected in X months': 1,
 'f you called a client to update them on their payment status and they were aggressive or rude to you, how would you handle it': 1,
 'an you think of a few common excuses that debtors make? How would you respond to each of them': 1,
 'hat collection methods do you know? Which technique do you think would be more effective for our company/clients': 1,
 'ow do you keep historical data for each account': 1,
 'ow often do you follow up with delinquent account holders and what’s your approach when you get in touch': 1,
 'ention three things a Collection Specialist should avoid saying when speaking to a debtor': 1,
 'ow do you keep track of all payments': 1,
 'hat software do you use to increase your productivity': 

In this step we order the dictionary of questions in descending order to check the "most repeated" questions:

In [19]:
{k: v for k, v in sorted(questions_freq.items(), key=lambda item: item[1],reverse=True)}

{'ow would you rate your computer skills': 5,
 'ow do you ensure accuracy in routine tasks such as processing expenses and preparing reports': 4,
 'hat constitutes a good unit test and what a functional one': 4,
 'hat accounting software have you used': 3,
 'ow familiar are you with SFAS (Statement of Financial Accounting Standards)': 3,
 'ow fast can you type': 3,
 'escribe your experience with calendar management': 3,
 'hat tools do you use for linting, debugging and profiling': 3,
 'hat do we mean when we say that a certain Lambda expression forms a closure': 3,
 'hat tools & practices would you consider necessary for a Continuous Delivery solution': 3,
 'o arguments in Java get passed by reference or by value': 3,
 'ow are Runtime exceptions different from Checked exceptions': 3,
 'ervlets 3.0 introduced async support. Describe a use case for it': 3,
 'ow do you troubleshoot a crashing application': 3,
 'hat’s your current occupation/What are you currently working on': 3,
 'ake us 

In [20]:
count=0
repetitions=0
uniques=0
rep=[]
for key,value in questions_freq.items():
    if value>1:
        count+=1
        repetitions+=value
        rep.append(key)
    else:
        uniques+=1
print("There are",count,"questions with at least 2 appearances")
print("There are",uniques,"unique questions")
print("There are a total of ",(uniques+repetitions),"questions appearances")


There are 169 questions with at least 2 appearances
There are 3972 unique questions
There are a total of  4342 questions appearances


In [21]:
count=0
rep=[]
for key,value in questions_freq.items():
        count+=1
        rep.append(key)
print(count)
len(rep)

4141


4141

## 2. Feature Engineering

### 2.1 Grouping Keywords with job descriptions

In [22]:
df_keywords = pd.read_csv("keywords.txt")

Creating random list to check if the code works 

In [23]:
keywords = ['accounting', 'job', 'Job brief','payments','and']
#df_keywords.Keywords.values.tolist() #changing keywords into a list 
print(keywords)

['accounting', 'job', 'Job brief', 'payments', 'and']


In [24]:
key_final = df_keywords.stack().tolist()
#key_final

In [25]:
df["keyword"] = df["job description"].str.findall('|'.join(keywords)).apply(set).str.join(', ')
df.head()

title  \
0        Collection Specialist   
1              Billing Analyst   
2               Tax Accountant   
3  Accounts Receivable Manager   
4              Cost Accountant   

                                     job description  \
0  <h4>Job brief</h4><p>We are looking for a comp...   
1  <h4>Job brief</h4><p>We are looking for a Bill...   
2  <h4>Job brief</h4><p>We are looking for a Tax ...   
3  <h4>Job brief</h4><p>We are looking for an Acc...   
4  <h4>Job brief</h4><p>We are looking for a Cost...   

                                           questions  number_of_questions  \
0  [['A debtor has missed six months of payments....                   16   
1  [['How would you create a report on debts from...                   14   
2  [['What would you do if you received a B-Notic...                   14   
3  [['What would be the first thing you’d check i...                   17   
4  [['If you could use only Excel, how would you ...                   15   

                                     keyword  
0                   and, payments, Job brief  
1            accounting, and, Job brief, job  
2  accounting, and, payments, Job brief, job  
3  accounting, and, payments, Job brief, job  
4                 accounting, and, Job brief

Once we have the keywords that match the job description, we create a list out of them for further processing:

In [26]:
df["keyword"]=df["keyword"].str.split(", ")
df.head()

title  \
0        Collection Specialist   
1              Billing Analyst   
2               Tax Accountant   
3  Accounts Receivable Manager   
4              Cost Accountant   

                                     job description  \
0  <h4>Job brief</h4><p>We are looking for a comp...   
1  <h4>Job brief</h4><p>We are looking for a Bill...   
2  <h4>Job brief</h4><p>We are looking for a Tax ...   
3  <h4>Job brief</h4><p>We are looking for an Acc...   
4  <h4>Job brief</h4><p>We are looking for a Cost...   

                                           questions  number_of_questions  \
0  [['A debtor has missed six months of payments....                   16   
1  [['How would you create a report on debts from...                   14   
2  [['What would you do if you received a B-Notic...                   14   
3  [['What would be the first thing you’d check i...                   17   
4  [['If you could use only Excel, how would you ...                   15   

                                       keyword  
0                   [and, payments, Job brief]  
1            [accounting, and, Job brief, job]  
2  [accounting, and, payments, Job brief, job]  
3  [accounting, and, payments, Job brief, job]  
4                 [accounting, and, Job brief]

### 2.2 Create a counter of keywords per question

In [63]:
column_list=[]
question_key_count={}
for role in df["questions"]:
    new_dict={}
    for question in role:
        new_dict[question]=0
    column_list.append(new_dict)
#column_list.append(question_key_count)
print(column_list)
print(len(column_list))

[{"['A debtor has missed six months of payments. When you call the individual, they bring up financial difficulties and get emotional. How do you handle this": 0, 'hat payback plan would you recommend for a long-term client whose payments are due to be collected in X months': 0, 'f you called a client to update them on their payment status and they were aggressive or rude to you, how would you handle it': 0, 'an you think of a few common excuses that debtors make? How would you respond to each of them': 0, 'hat collection methods do you know? Which technique do you think would be more effective for our company/clients': 0, 'ow do you keep historical data for each account': 0, 'ow often do you follow up with delinquent account holders and what’s your approach when you get in touch': 0, 'ention three things a Collection Specialist should avoid saying when speaking to a debtor': 0, 'ow do you keep track of all payments': 0, 'hat software do you use to increase your productivity': 0, 'hat 

In [62]:
column_list=[]
question_key_count={}
for role in df["questions"]:
    new_dict={}
    for question in role:
        new_dict[question]=0
    column_list.append(new_dict)

i=0
keyword_count=0
for role_dict in column_list:  
    for question in role_dict.keys():
        for keyword in df["keyword"][i]: 
            if keyword in question:  
                keyword_count+=1
        role_dict[question]=keyword_count
        keyword_count=0
    i+=1
            
print(column_list)
print(len(column_list))

[{"['A debtor has missed six months of payments. When you call the individual, they bring up financial difficulties and get emotional. How do you handle this": 2, 'hat payback plan would you recommend for a long-term client whose payments are due to be collected in X months': 1, 'f you called a client to update them on their payment status and they were aggressive or rude to you, how would you handle it': 1, 'an you think of a few common excuses that debtors make? How would you respond to each of them': 0, 'hat collection methods do you know? Which technique do you think would be more effective for our company/clients': 0, 'ow do you keep historical data for each account': 0, 'ow often do you follow up with delinquent account holders and what’s your approach when you get in touch': 1, 'ention three things a Collection Specialist should avoid saying when speaking to a debtor': 0, 'ow do you keep track of all payments': 1, 'hat software do you use to increase your productivity': 0, 'hat 

In [60]:
df["keywords_per_question"]=column_list
df.head()

title  \
0        Collection Specialist   
1              Billing Analyst   
2               Tax Accountant   
3  Accounts Receivable Manager   
4              Cost Accountant   

                                     job description  \
0  <h4>Job brief</h4><p>We are looking for a comp...   
1  <h4>Job brief</h4><p>We are looking for a Bill...   
2  <h4>Job brief</h4><p>We are looking for a Tax ...   
3  <h4>Job brief</h4><p>We are looking for an Acc...   
4  <h4>Job brief</h4><p>We are looking for a Cost...   

                                           questions  number_of_questions  \
0  [['A debtor has missed six months of payments....                   16   
1  [['How would you create a report on debts from...                   14   
2  [['What would you do if you received a B-Notic...                   14   
3  [['What would be the first thing you’d check i...                   17   
4  [['If you could use only Excel, how would you ...                   15   

                                       keyword  \
0                   [and, payments, Job brief]   
1            [accounting, and, Job brief, job]   
2  [accounting, and, payments, Job brief, job]   
3  [accounting, and, payments, Job brief, job]   
4                 [accounting, and, Job brief]   

                               keywords_per_question  
0  {'['A debtor has missed six months of payments...  
1  {'['How would you create a report on debts fro...  
2  {'['What would you do if you received a B-Noti...  
3  {'['What would be the first thing you’d check ...  
4  {'['If you could use only Excel, how would you...

In [61]:
df["keywords_per_question"][0]

{"['A debtor has missed six months of payments. When you call the individual, they bring up financial difficulties and get emotional. How do you handle this": 2,
 'hat payback plan would you recommend for a long-term client whose payments are due to be collected in X months': 1,
 'f you called a client to update them on their payment status and they were aggressive or rude to you, how would you handle it': 1,
 'an you think of a few common excuses that debtors make? How would you respond to each of them': 0,
 'hat collection methods do you know? Which technique do you think would be more effective for our company/clients': 0,
 'ow do you keep historical data for each account': 0,
 'ow often do you follow up with delinquent account holders and what’s your approach when you get in touch': 1,
 'ention three things a Collection Specialist should avoid saying when speaking to a debtor': 0,
 'ow do you keep track of all payments': 1,
 'hat software do you use to increase your productivity': 

In [58]:
string_list=["Hello world, my name is JD and I'm running a little test to see what happens","I love playing a basketball match"]
list_of_words=["world", "JD","see","basketball", "rugby", "match"]
word_count=0
words_per_string={}

for string in string_list:
    words_per_string[string]=0
    for word in list_of_words:
        if word in string:
            words_per_string[string]+=1
print(words_per_string)

{"Hello world, my name is JD and I'm running a little test to see what happens": 2, 'I love playing a basketball match': 2}
